compiler/fir/checkers/checkers.native/src/org/jetbrains/kotlin/fir/analysis/native/checkers/FirNativeObjCRefinementChecker.kt

```diff
@@ -0,0 +1,242 @@
+            if (superIsRefinedInSwift) isRefinedInSwift = true else supersNotRefinedInSwift.add(overriddenMemberSymbol)
+        }
+        if (isHiddenFromObjC && supersNotHiddenFromObjC.isNotEmpty()) {
+            reporter.reportIncompatibleOverride(memberSymbol.fir, objCAnnotations, supersNotHiddenFromObjC, context)
+        }
+        if (isRefinedInSwift && supersNotRefinedInSwift.isNotEmpty()) {
+            reporter.reportIncompatibleOverride(memberSymbol.fir, swiftAnnotations, supersNotRefinedInSwift, context)
+        }
+    }
+
+    private fun FirCallableSymbol<*>.inheritsRefinedAnnotations(context: CheckerContext): Pair<Boolean, Boolean> {
+        val (hasObjC, hasSwift) = fir.hasRefinedAnnotations(context.session)
+        if (hasObjC && hasSwift) return true to true
+        // Note: `checkMember` requires all overridden symbols to be either refined or not refined.
+        val overriddenMemberSymbol = fir.containingClass()
+            ?.toFirRegularClass(context.session)
+            ?.unsubstitutedScope(context)
+            ?.retrieveDirectOverriddenOf(this)
+            ?.firstOrNull()
+            ?: return hasObjC to hasSwift
+        val (inheritsObjC, inheritsSwift) = overriddenMemberSymbol.inheritsRefinedAnnotations(context)
+        return (hasObjC || inheritsObjC) to (hasSwift || inheritsSwift)
+    }
+
+    private fun FirDeclaration.hasRefinedAnnotations(session: FirSession): Pair<Boolean, Boolean> {
+        var hasObjC = false
+        var hasSwift = false
+        for (annotation in annotations) {
+            val metaAnnotations = annotation.getMetaAnnotations(session)
+            for (metaAnnotation in metaAnnotations) {
+                when (metaAnnotation.toAnnotationClassId()) {
+                    hidesFromObjCClassId -> {
+                        hasObjC = true
+                        break
+                    }
+                    refinesInSwiftClassId -> {
+                        hasSwift = true
+                        break
+                    }
+                }
+            }
+            if (hasObjC && hasSwift) return true to true
+        }
+        return hasObjC to hasSwift
+    }
+
+    private fun FirAnnotation.getMetaAnnotations(session: FirSession): Iterable<FirAnnotation> =
+        coneClassLikeType?.lookupTag?.toSymbol(session)?.fir?.annotations.orEmpty()
+
+    private fun FirTypeScope.retrieveDirectOverriddenOf(memberSymbol: FirCallableSymbol<*>): List<FirCallableSymbol<*>> {
+        return when (memberSymbol) {
+            is FirNamedFunctionSymbol -> {
+                processFunctionsByName(memberSymbol.name) {}
+                getDirectOverriddenFunctions(memberSymbol)
+            }
+            is FirPropertySymbol -> {
+                processPropertiesByName(memberSymbol.name) {}
+                getDirectOverriddenProperties(memberSymbol)
+            }
+            else -> throw IllegalArgumentException("unexpected member kind $memberSymbol")
```

Comment:
Style: it's better to use `error("message")` instead of `throw IllegalArgumentException("message")`